[pytorch] Early return in nn.EmbeddingBag when weight is empty

xw285cornell opened the pull request:
Summary: When `num_samples == 0`, grid becomes zero. Although CUDA just silently proceeds, `cudaGetLastError()` will complain about the `Error: invalid configuration argument`. So it's actually failing in some future places that becomes really hard to debug.

Differential Revision: D24409874

serhaty commented:
hey @xw285cornell , lgtm. can we also add a unit test case to cover this? Thanks!

xw285cornell commented:
I added unit test in the internal version to unblock some work. I'll do a follow up on the OSS unit test. 

@ngimel how does aten treats empty input tensor in general? Launch a kernel w/ grid=0 seems to be ok with CUDA except the error from the cudaGetLastError(). Not sure if this problem can happen somewhere else too.

ngimel commented:
Yeah, please add unittest. Thanks for finding this, @r-barnes is now working on systematically checking all launches #46474.
There are usually early returns in aten for 0-element tensors, to not hit "invalid configuration argument" error.

r-barnes commented:
@xw285cornell : An early return seems best to me since the return carries semantics that the kernel launch does not (i.e. makes it clear what's going on).

xw285cornell committed:
[pytorch] Early return in nn.EmbeddingBag when weight is empty (#46572)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46572

When `num_samples == 0`, grid becomes zero. Although CUDA just silently proceeds, `cudaGetLastError()` will complain about the `Error: invalid configuration argument`. So it's actually failing in some future places that becomes really hard to debug.

Reviewed By: jianyuh

Differential Revision: D24409874

fbshipit-source-id: c35412c47835cb0770745ef0615215c0e47c2f50

xw285cornell commented:
Unit test added, @ngimel @serhaty  could you take a look?
